node/runtime/pangolin/src/pallets/bridge_dispatch.rs

```diff
@@ -3,19 +3,80 @@ pub use pallet_bridge_dispatch::{
 };
 
 // --- paritytech ---
-use frame_support::traits::Everything;
+use frame_support::traits::OriginTrait;
+use sp_runtime::transaction_validity::{InvalidTransaction, TransactionValidityError};
 // --- darwinia-network ---
 use crate::*;
+use bp_message_dispatch::{CallValidate, IntoDispatchOrigin as IntoDispatchOriginT};
 use bp_messages::{LaneId, MessageNonce};
+use darwinia_ethereum::{RawOrigin, Transaction};
+use darwinia_evm::AccountBasic;
+use darwinia_support::evm::{DeriveEthereumAddress, DeriveSubstrateAddress};
 use pallet_bridge_dispatch::Config;
 
+pub struct CallValidator;
+impl CallValidate<bp_pangolin::AccountId, Origin, Call> for CallValidator {
+	fn pre_dispatch(
+		relayer_account: &bp_pangolin::AccountId,
+		origin: &Origin,
+		call: &Call,
+	) -> Result<(), TransactionValidityError> {
+		match call {
+			// Note: Only supprt Ethereum::transact(LegacyTransaction)
+			Call::Ethereum(darwinia_ethereum::Call::transact { transaction: tx }) => {
+				match origin.caller() {
+					OriginCaller::Ethereum(RawOrigin::EthereumTransaction(id)) => {
+						match tx {
+							Transaction::Legacy(t) => {
+								let fee = t.gas_limit.saturating_mul(t.gas_limit);
+								let total_payment = fee.saturating_add(t.value);
```

Comment:
This could be an attack to relayer, to run out of their balances.

Solution:

Do not support cases `t.value > 0` for now.

Reply:
TODO here.

Add checking for gas_price.

Reply:
done